source/agora/node/EnrollmentManager.d

```diff
@@ -0,0 +1,426 @@
+
+    ***************************************************************************/
+
+    public int opApply (scope int delegate(Hash hash, EnrollmentData enroll) dg) @trusted
+    {
+        auto results = this.db.execute("SELECT key, val FROM validator_set");
+
+        foreach (row; results)
+        {
+            auto hash = *cast(Hash*)row.peek!(ubyte[])(0).ptr;
+            auto enroll = deserializeFull!EnrollmentData(row.peek!(ubyte[])(1));
+
+            // break
+            if (auto ret = dg(hash, enroll))
+                return ret;
+        }
+
+        return 0;
+    }
+
+    /***************************************************************************
+
+        Remove the enrollment data with the given key from the set
+
+        Params:
+            enroll = the enrollment data to remove
+
+    ***************************************************************************/
+
+    public void remove (const ref EnrollmentData enroll) @trusted
+    {
+        auto hash = hashFull(enroll.K);
+        this.db.execute("DELETE FROM validator_set WHERE key = ?", hash[]);
+    }
+
+    /***************************************************************************
+
+        Take the specified number of enrollment data and remove them from the set.
+
+        Params:
+            count = how many enrollment data to take from the set
+
+        Returns:
+            an array of 'count' enrollment data
+
+    ***************************************************************************/
+
+    version (unittest) public EnrollmentData[] take (size_t count) @safe
+    {
+        const len_prev = this.length();
+        assert(len_prev >= count);
+        EnrollmentData[] enrolls;
+
+        foreach (hash, enroll; this)
+        {
+            enrolls ~= enroll;
+            if (enrolls.length == count)
+                break;
+        }
+
+        enrolls.each!(enroll => this.remove(enroll));
+        assert(this.length() == len_prev - count);
+        return enrolls;
+    }
+
+    /***************************************************************************
+
+        Makes a data object for enrollment process
+
+        Returns:
+            a EnrollmentData object
+
+    ***************************************************************************/
+
+    public EnrollmentData createEnrollmentData () @trusted
+    {
+        this.data = new EnrollmentData();
+
+        // public key
+        this.data.K = this.node_keypair.address;
+
+        // cycle length
+        this.data.n = 1008; // freezing period / 2
+
+        // random seed
+        this.data.X = hashFull(this.node_keypair.secret.data);
+        foreach (i; 0 .. this.data.n-1)
+            this.data.X = hashFull(this.data.X);
+
+        // signature noise
+        this.schnorr_pair = Pair.random();
+        this.data.R = this.schnorr_pair.V;
+
+        // signature
+        Hash hash = hashFull(*this.data);
+        this.data.S = sign(this.schnorr_pair, hash[]);
```

Comment:
I don't think this is right. Here you are proving ownership of `R` by using `r` as a signing key, and randomly generating the actual R in the signature (see the `sign` method overload used here, it randomly generates R). But, this doesn't prove that you know `k` of `K`.

Instead I suggest signing it like so:

```D
auto pk = Scalar(this.node_keypair.seed[]);
this.data.S = sign(kp, kp.toPoint(), this.schnorr_pair.V, this.schnorr_pair.v, hash[]);
```